### ISSUE TITLE
DL3066 doesn't expand variables resulting in false positives

### ISSUE THREAD

**Issue by nickjj** [hadolint/hadolint#1222]
Originally I had `USER python` in my Dockerfile in which Hadolint v2.15.0 now throws this:

```
-:16 DL3066 info: Non-numeric user-id may not be resolvable by host system
```

But my Dockerfile has this where the APP_UID is configurable at build time:

```dockerfile
ARG APP_UID=1000
ARG APP_GID=1000

RUN apt-get update \
  && apt-get install -y --no-install-recommends build-essential curl libpq-dev \
  && rm -rf /var/lib/apt/lists/* /usr/share/doc /usr/share/man \
  && apt-get clean \
  && groupadd -g "${APP_GID}" python \
  && useradd --create-home --no-log-init -u "${APP_UID}" -g "${APP_GID}" python \
  && chown python:python -R /app

COPY --from=ghcr.io/astral-sh/uv:0.8.17 /uv /uvx /usr/local/bin/

USER ${APP_UID}
```

However, Hadolint still throws:

```
-:16 DL3066 info: Non-numeric user-id may not be resolvable by host system
```

Of course I can hard code `USER 1000` but that UID may not exist if the image is built with a non-1000 value. The original `USER python` approach technically is a suitable solution here I believe, but in this case, Hadolint is also failing to resolve the variable and also failing to account that `useradd` is creating a correctly mapped UID so referencing it by name is ok.

In this case the `python` user only needs to exist in the image / container (which it does), if you add a volume mount, it correctly happens using the UID:GID, the host does not need the `python` user for it to work.

**Comment by m-ildefons** [comment]
Thanks for reporting this.

You can work around this until it's fixed by disabling the rule. See the readme file for how to do that.

I'm curious: why do you need to make the user's UID configurable at build time? Wouldn't it be just fine to just put a static `1000` instead of `${APP_UID}` and `${APP_GID}` respectively?

**Comment by nickjj** [comment]
Thanks, yep I ignored it.

Usually when you deploy to a Linux server, it would be with the first user (uid:gid 1000) but sometimes you don't have control over this and it might be 1001 or some other number. If you volume mount files such as static files that you serve from nginx on the Docker host, then volume mounts need to line up to the correct UID on the host.

Having the above snippet lets you adjust this at build time for whatever hosting environment you have. I have it set up with a single `.env` file which gets used through Docker Compose.

It's a pattern I use on these example projects if you're looking for an end to end example to test the rule:

- https://github.com/nickjj/docker-flask-example
- https://github.com/nickjj/docker-django-example
- https://github.com/nickjj/docker-rails-example
- https://github.com/nickjj/docker-node-example

Here's a blog post with more details on this pattern: https://nickjanetakis.com/blog/running-docker-containers-as-a-non-root-user-with-a-custom-uid-and-gid

**Comment by m-ildefons** [comment]
For sharing static files from a Linux server, I'd make them world-readable (`chmod o+r`) and then bind-mounted the directory/volume into the container. Preferably with a read-only mount.
This way the UID in the container doesn't need to map to any UID on the host. This is also pretty much what the Nginx container image docs recommend.

You only need to match UIDs/GIDs when you need to _write_ to the volume (i.e. file uploads to your web app). But in that case, I would rather just `chown` the directory on the host system to match the UID/GID from the container than the other way around. Keep in mind, you don't need to actually have a user/group with that ID on the host.

The only other case I can think of is when your container image is executed in an environment where you don't have control over such things. I.e. a cloud. But then plain file storage for uploads is also not really the way to go, since in a cloud you would most likely be better off using the cloud's object storage service.

Regardless, I've added a fix.
I'm also making a new release asap, since there was a problem with the MacOS builds for v2.15.0.

**Comment by FranPavelic-SB** [comment]
This warning is also missing a wiki page :)

**Comment by nickjj** [comment]
> You only need to match UIDs/GIDs when you need to write to the volume (i.e. file uploads to your web app).

All of my above projects write because the images build assets with Tailwind and esbuild which then get mounted at runtime. The container's entrypoint copies the assets from the image to the mounted directory (a write operation from the host's POV). Having the project self contained where you can up it and be done with it, instead of needing to provision a special directory ahead of time is nice.

Thanks for addressing it, I'll keep an eye out for the next release.

**Comment by nickjj** [comment]
Btw, I checked the PR and I think it will still be a false positive if you put `USER python` but before setting that you've used `useradd` to create a user with that name. It's a false positive because the lint error is "Non-numeric user-id may not be resolvable by host system", but it will be resolvable by the host system since the volume mount happens over the UID not the name.

I know through static analysis that might be tough to address, but in the end I may end up keeping the rule ignored since from a readability POV, keeping `USER python` in the `Dockerfile` is easier to see what's happening.

**Comment by m-ildefons** [comment]
DL3066 was added because if you run an unprivileged container in Kubernetes by just setting `.spec.securityContext.runAsNonRoot=true` for your Pod, then the Pod may fail to start. This is explained in more detail in the issue linked from the PR: https://github.com/hadolint/hadolint/issues/1202
I decided to add this rule, since Kubernetes is so popular and running a container as non-root user is considered a good practice. Therefore Hadolint should help make it easy to build containers that just work for that use case.
If your use case doesn't involve Kubernetes, or you set `.spec.securityContext.runAsUser`, then this rule isn't particularly helpful. It's completely ok to disable the rule if it doesn't apply to your use case.

I'll close this issue, since v2.15.1 will not warn on `USER ${APP_UID}` if `APP_UID` has been defined as `ARG`. But thanks for opening this issue and the productive discussion. This really helps improve Hadolint.

**Comment by acdha** [comment]
> I'll close this issue, since v2.15.1 will not warn on `USER ${APP_UID}` if `APP_UID` has been defined as `ARG`. But thanks for opening this issue and the productive discussion. This really helps improve Hadolint.

Since this breaks almost everyone following container best-practices would it be possible to either have the error message note this or link to some more detail about it? The pattern of parsing `useradd` / `adduser` calls would also fix it but that'd take on a lot of extremely tedious shell parsing rules and dealing with the difference between the various implementations of those utilities so I think a warning makes sense as long as it's clear what the user should do to resolve it.

This also feels like there might be a value in having some concept of a timed ignore where you might want to ignore DL3066 for 3 months but still display it so developers can see and fix it before it starts failing. A lot of security scanners have some kind of “expires at” concept but for a linter it feels like the semantics should be “display as an error but do not cause a non-zero exit”.
